vllm/model_executor/models/paligemma.py

```diff
@@ -47,95 +50,170 @@ class PaliGemmaImageEmbeddingInputs(TypedDict):
+    ) -> Mapping[str, MultiModalFieldConfig]:
+        return dict(pixel_values=MultiModalFieldConfig.batched("image"))
 
-    def forward(self, image_features: torch.Tensor) -> torch.Tensor:
-        hidden_states = self.linear(image_features)
-        return hidden_states
+    def _apply_hf_processor_tokens_only(
+        self,
+        prompt_tokens: list[int],
+    ) -> list[int]:
+        tokenizer = self.info.get_tokenizer()
+        # Adding <bos> token (tokenizer.bos_token_id) to match with _call_hf_processor
+        if len(prompt_tokens) == 0:
+            tokens_with_bos = [tokenizer.bos_token_id]
+        elif prompt_tokens[0] != tokenizer.bos_token_id:
+            tokens_with_bos = [tokenizer.bos_token_id] + prompt_tokens
+        else:
+            tokens_with_bos = prompt_tokens
+        return tokens_with_bos
 
+    def _get_prompt_replacements(
+        self,
+        mm_items: MultiModalDataItems,
+        hf_processor_mm_kwargs: Mapping[str, object],
+        out_mm_kwargs: MultiModalKwargs,
+    ) -> list[PromptReplacement]:
+        hf_config = self.info.get_hf_config()
+        image_token_id = hf_config.image_token_index
+
+        tokenizer = self.info.get_tokenizer()
+        num_image_tokens = self.info.get_num_image_tokens()
+        image_tokens = [image_token_id] * num_image_tokens
+
+        bos_token_id = tokenizer.bos_token_id
+        assert isinstance(bos_token_id, int)
+
+        # Adding <bos> token at the beginning based on add_bos_token variable.
+        # Always adding <bos> token at the end of the images tokens and before the text prompt
+        # according to Paligemma's format
+        if tokenizer.add_bos_token:
+            replacement_tokens = [bos_token_id] + image_tokens + [bos_token_id]
+        else:
+            replacement_tokens = image_tokens + [bos_token_id]
+
+        return [
+            PromptReplacement(
+                modality="image",
+                target=[bos_token_id],
+                replacement=PromptReplacementDetails(
+                    full=replacement_tokens,
+                    features=image_tokens,
+                ),
+            )
+        ]
 
-@MULTIMODAL_REGISTRY.register_image_input_mapper()
-@MULTIMODAL_REGISTRY.register_max_image_tokens(get_max_paligemma_image_tokens)
-@INPUT_REGISTRY.register_dummy_data(dummy_data_for_paligemma)
-@INPUT_REGISTRY.register_input_processor(input_processor_for_paligemma)
+    def apply(
+        self,
+        prompt: Union[str, list[int]],
+        mm_data: MultiModalDataDict,
+        hf_processor_mm_kwargs: Mapping[str, object],
+    ) -> MultiModalInputs:
+        multimodelinputs = super().apply(prompt, mm_data,
+                                         hf_processor_mm_kwargs)
+        # Force to add newline at the end of prompt according to paligemma's format
+        prompt_token_ids = multimodelinputs["prompt_token_ids"]
+        if len(prompt_token_ids) and prompt_token_ids[-1] != 108:
```

Comment:
```suggestion
        if len(prompt_token_ids) and prompt_token_ids[-1] != 108:
```

Let's avoid using a magic number here